gh-112970: Detect and use closefrom() on platforms where it is available

thesamesam opened the pull request:
glibc-2.34 implements closefrom(3) using the same semantics as on BSD. Check for closefrom in configure and use the check result in fileutils.c rather than hardcoding a FreeBSD check (which was wrong for other BSDs anyway).


<!-- gh-issue-number: gh-112970 -->
* Issue: gh-112970
<!-- /gh-issue-number -->

thesamesam committed:
gh-112970: Wire up closefrom for Linux

glibc-2.34 implements closefrom(3) using the same semantics as on BSD. Check
for closefrom in configure and use the check result in fileutils.c rather than
hardcoding a FreeBSD check (which was wrong for other BSDs anyway).

Signed-off-by: Sam James <sam@gentoo.org>

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

erlend-aasland commented on Python/fileutils.c:
```diff
@@ -2878,9 +2878,9 @@ _Py_GetLocaleconvNumeric(struct lconv *lc,
  */
-#ifdef __FreeBSD__
+#ifdef HAVE_CLOSEFROM
 #  define USE_CLOSEFROM
-#endif /* __FreeBSD__ */
+#endif /* HAVE_CLOSEFROM */
```
Another option is to remove the `__FreeBSD__` guard and `s/USE_CLOSEFROM/HAVE_CLOSEFROM/g` in this file. This is a smaller diff, though. I'm fine with either.

thesamesam replied:
Yeah, fair point. I kept it this way for consistency with the `fdwalk` check, but if desired, I can change it. My preference is probably to keep it as-is though in the absence of anyone feeling strongly.

erlend-aasland replied:
This is fine.

erlend-aasland commented on Misc/NEWS.d/next/Library/2023-12-11-16-13-15.gh-issue-112970.87jmKP.rst:
```diff
@@ -0,0 +1 @@
+Use closefrom on Linux where available (e.g. glibc-2.34), rather than only FreeBSD.
```
```suggestion
Use :c:func:`!closefrom` on Linux where available (e.g. glibc-2.34), rather than only FreeBSD.
```

erlend-aasland reviewed (approved):
Thanks, this makes sense.

thesamesam committed:
Fix NEWS item

Signed-off-by: Sam James <sam@gentoo.org>

thesamesam committed:
Explicitly ignore closefrom return value

There are two variants of closefrom, one returns void, but another returns int,
so let's cast the return value to void to explicitly discard it to avoid -Wunused-result
warnings.

Signed-off-by: Sam James <sam@gentoo.org>
